### ISSUE TITLE
capa-controller-manager continously creates new launchtemplate versions

### ISSUE THREAD

**Issue by matlemb** [kubernetes-sigs/cluster-api-provider-aws#3372]
/kind bug

**What steps did you take and what happened:**
`capa-controller-manager` continously creates new launchtemplate versions even when manual changes are done to the `AWSMachinePool` resource. At first i thought this might be an tag ordering issue because i applied additional tags via `AWSMachinePool.spec.additionalTags` which where ordered differently on each version when checking the launchtemplate in AWS UI, but after removing the labels and recreating the MachinePool and AWSMachinePool it still happens. Also tried not specifing `AWSMachinePool.spec.awsLaunchTemplate.versionNumber` which did not improve anything.

## MachinePool + AWSMachinePool YAML
```YAML
kind: MachinePool
metadata:
  name: my-aws-workload-cluster-mp-0
  namespace: my-aws-workload-cluster
spec:
  clusterName: my-aws-workload-cluster
  replicas: 3
  template:
    spec:
      bootstrap:
        configRef:
          apiVersion: bootstrap.cluster.x-k8s.io/v1beta1
          kind: KubeadmConfig
          name: my-aws-workload-cluster-md-0
      clusterName: my-aws-workload-cluster
      infrastructureRef:
        apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
        kind: AWSMachinePool
        name: my-aws-workload-cluster-mp-0
      version: v1.23.3
---
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: AWSMachinePool
metadata:
  name: my-aws-workload-cluster-mp-0
  namespace: my-aws-workload-cluster
spec:
  additionalTags:
          #aws-node-termination-handler/managed: ""
          #k8s.io/cluster-autoscaler/enabled: ""
          #k8s.io/cluster-autoscaler/my-aws-workload-cluster: ""
  minSize: 1
  maxSize: 10
  capacityRebalance: true
  availabilityZones:
    - eu-central-1a
    - eu-central-1b
    - eu-central-1c
  awsLaunchTemplate:
    iamInstanceProfile: nodes.cluster-api-provider-aws.sigs.k8s.io
    imageLookupBaseOS: ubuntu-20.04
    instanceType: "t3a.medium"
    sshKeyName: "capa"
    rootVolume:
      size: 40
    versionNumber: 1
  mixedInstancesPolicy:
    instancesDistribution:
      onDemandAllocationStrategy: prioritized
      onDemandBaseCapacity: 0
      onDemandPercentageAboveBaseCapacity: 0
      spotAllocationStrategy: lowest-price
    overrides:
    - instanceType: "t3.small"
    - instanceType: "t3.medium"
    - instanceType: "t3.large"
    - instanceType: "t3.xlarge"
    - instanceType: "t3a.small"
    - instanceType: "t3a.medium"
    - instanceType: "t3a.large"
    - instanceType: "t3a.xlarge"
  subnets:
  - id: subnet-aaa
  - id: subnet-bbb
  - id: subnet-ccc
```

## capa controller logs:
```
I0401 12:12:34.324146       1 awsmachinepool_controller.go:406]  "msg"="no existing launch template found, creating"  
I0401 12:12:36.854424       1 awsmachinepool_controller.go:460]  "msg"="creating new version for launch template"  "existing"={"name":"aws-workload-001-1379-r-01-mp-0","iamInstanceProfile":"nodes.cluster-api-provider-aws.sigs.k8s.io","ami":{"id":"ami-0a2e26e8ab71b30ac"},"instanceType":"t3a.medium","sshKeyName":"capa","versionNumber":1,"additionalSecurityGroups":[{"id":"sg-0ca464abfefa011dc"},{"id":"sg-0e156080d2b752180"}]} "incoming"={"iamInstanceProfile":"nodes.cluster-api-provider-aws.sigs.k8s.io","ami":{},"imageLookupBaseOS":"ubuntu-20.04","instanceType":"t3a.medium","rootVolume":{"deviceName":"/dev/sda1","size":40},"sshKeyName":"capa","versionNumber":1}
I0401 12:21:33.727728       1 awsmachinepool_controller.go:460]  "msg"="creating new version for launch template"  "existing"={"name":"aws-workload-001-1379-r-01-mp-0","iamInstanceProfile":"nodes.cluster-api-provider-aws.sigs.k8s.io","ami":{"id":"ami-0a2e26e8ab71b30ac"},"instanceType":"t3a.medium","sshKeyName":"capa","versionNumber":2,"additionalSecurityGroups":[{"id":"sg-0ca464abfefa011dc"},{"id":"sg-0e156080d2b752180"}]} "incoming"={"iamInstanceProfile":"nodes.cluster-api-provider-aws.sigs.k8s.io","ami":{},"imageLookupBaseOS":"ubuntu-20.04","instanceType":"t3a.medium","rootVolume":{"deviceName":"/dev/sda1","size":40},"sshKeyName":"capa","versionNumber":1}
I0401 12:28:49.261315       1 awsmachinepool_controller.go:460]  "msg"="creating new version for launch template"  "existing"={"name":"aws-workload-001-1379-r-01-mp-0","iamInstanceProfile":"nodes.cluster-api-provider-aws.sigs.k8s.io","ami":{"id":"ami-0a2e26e8ab71b30ac"},"instanceType":"t3a.medium","sshKeyName":"capa","versionNumber":3,"additionalSecurityGroups":[{"id":"sg-0ca464abfefa011dc"},{"id":"sg-0e156080d2b752180"}]} "incoming"={"iamInstanceProfile":"nodes.cluster-api-provider-aws.sigs.k8s.io","ami":{},"imageLookupBaseOS":"ubuntu-20.04","instanceType":"t3a.medium","rootVolume":{"deviceName":"/dev/sda1","size":40},"sshKeyName":"capa","versionNumber":1}
I0401 12:38:43.060429       1 awsmachinepool_controller.go:460]  "msg"="creating new version for launch template"  "existing"={"name":"aws-workload-001-1379-r-01-mp-0","iamInstanceProfile":"nodes.cluster-api-provider-aws.sigs.k8s.io","ami":{"id":"ami-0a2e26e8ab71b30ac"},"instanceType":"t3a.medium","sshKeyName":"capa","versionNumber":4,"additionalSecurityGroups":[{"id":"sg-0ca464abfefa011dc"},{"id":"sg-0e156080d2b752180"}]} "incoming"={"iamInstanceProfile":"nodes.cluster-api-provider-aws.sigs.k8s.io","ami":{},"imageLookupBaseOS":"ubuntu-20.04","instanceType":"t3a.medium","rootVolume":{"deviceName":"/dev/sda1","size":40},"sshKeyName":"capa","versionNumber":1}
I0401 12:45:45.766494       1 awsmachinepool_controller.go:460]  "msg"="creating new version for launch template"  "existing"={"name":"aws-workload-001-1379-r-01-mp-0","iamInstanceProfile":"nodes.cluster-api-provider-aws.sigs.k8s.io","ami":{"id":"ami-0a2e26e8ab71b30ac"},"instanceType":"t3a.medium","sshKeyName":"capa","versionNumber":5,"additionalSecurityGroups":[{"id":"sg-0ca464abfefa011dc"},{"id":"sg-0e156080d2b752180"}]} "incoming"={"iamInstanceProfile":"nodes.cluster-api-provider-aws.sigs.k8s.io","ami":{},"imageLookupBaseOS":"ubuntu-20.04","instanceType":"t3a.medium","rootVolume":{"deviceName":"/dev/sda1","size":40},"sshKeyName":"capa","versionNumber":1}
I0401 12:48:36.892370       1 awsmachinepool_controller.go:460]  "msg"="creating new version for launch template"  "existing"={"name":"aws-workload-001-1379-r-01-mp-0","iamInstanceProfile":"nodes.cluster-api-provider-aws.sigs.k8s.io","ami":{"id":"ami-0a2e26e8ab71b30ac"},"instanceType":"t3a.medium","sshKeyName":"capa","versionNumber":6,"additionalSecurityGroups":[{"id":"sg-0ca464abfefa011dc"},{"id":"sg-0e156080d2b752180"}]} "incoming"={"iamInstanceProfile":"nodes.cluster-api-provider-aws.sigs.k8s.io","ami":{},"imageLookupBaseOS":"ubuntu-20.04","instanceType":"t3a.medium","rootVolume":{"deviceName":"/dev/sda1","size":40},"sshKeyName":"capa","versionNumber":1}
I0401 12:58:30.755570       1 awsmachinepool_controller.go:460]  "msg"="creating new version for launch template"  "existing"={"name":"aws-workload-001-1379-r-01-mp-0","iamInstanceProfile":"nodes.cluster-api-provider-aws.sigs.k8s.io","ami":{"id":"ami-0a2e26e8ab71b30ac"},"instanceType":"t3a.medium","sshKeyName":"capa","versionNumber":7,"additionalSecurityGroups":[{"id":"sg-0ca464abfefa011dc"},{"id":"sg-0e156080d2b752180"}]} "incoming"={"iamInstanceProfile":"nodes.cluster-api-provider-aws.sigs.k8s.io","ami":{},"imageLookupBaseOS":"ubuntu-20.04","instanceType":"t3a.medium","rootVolume":{"deviceName":"/dev/sda1","size":40},"sshKeyName":"capa","versionNumber":1}
I0401 13:08:15.077124       1 awsmachinepool_controller.go:460]  "msg"="creating new version for launch template"  "existing"={"name":"aws-workload-001-1379-r-01-mp-0","iamInstanceProfile":"nodes.cluster-api-provider-aws.sigs.k8s.io","ami":{"id":"ami-0a2e26e8ab71b30ac"},"instanceType":"t3a.medium","sshKeyName":"capa","versionNumber":8,"additionalSecurityGroups":[{"id":"sg-0ca464abfefa011dc"},{"id":"sg-0e156080d2b752180"}]} "incoming"={"iamInstanceProfile":"nodes.cluster-api-provider-aws.sigs.k8s.io","ami":{},"imageLookupBaseOS":"ubuntu-20.04","instanceType":"t3a.medium","rootVolume":{"deviceName":"/dev/sda1","size":40},"sshKeyName":"capa","versionNumber":1}
I0401 13:17:35.439797       1 awsmachinepool_controller.go:460]  "msg"="creating new version for launch template"  "existing"={"name":"aws-workload-001-1379-r-01-mp-0","iamInstanceProfile":"nodes.cluster-api-provider-aws.sigs.k8s.io","ami":{"id":"ami-0a2e26e8ab71b30ac"},"instanceType":"t3a.medium","sshKeyName":"capa","versionNumber":9,"additionalSecurityGroups":[{"id":"sg-0ca464abfefa011dc"},{"id":"sg-0e156080d2b752180"}]} "incoming"={"iamInstanceProfile":"nodes.cluster-api-provider-aws.sigs.k8s.io","ami":{},"imageLookupBaseOS":"ubuntu-20.04","instanceType":"t3a.medium","rootVolume":{"deviceName":"/dev/sda1","size":40},"sshKeyName":"capa","versionNumber":1}

```

**What did you expect to happen:**
capa-controller-manager should not continously create new launchtemplate versions when no changes were made.

**Anything else you would like to add:**

**Environment:**

- Cluster-api-provider-aws version: 1.3.0
- Kubernetes version: (use `kubectl version`): 1.23.3
- OS (e.g. from `/etc/os-release`): ubuntu 20.04
<sub>Matthias Lembcke <[matthias.lembcke@mercedes-benz.com](mailto:matthias.lembcke@mercedes-benz.com)>, Daimler TSS GmbH ([Provider Information](https://github.com/mercedes-benz/daimler-foss/blob/master/PROVIDER_INFORMATION.md))</sub>

**Comment by sedefsavas** [comment]
This is a feature, added with https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/2354.
TL;DR Bootstrap token expires every 15 min, and when a new template is not created, during scale up, new machines fails to join the cluster. 
New launch templates are created on user data (bootstrap token) changes.

One todo left was cleaning up older launch templates. Will checkout if we have an issue for this.

**Comment by sedefsavas** [comment]
Found this fix on cleaning up the old templates: https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/2525

Continuous creation of LaunchTemplates are on purpose but are the old versions not being cleaned up in your tests? If so, this is a bug.

**Comment by matlemb** [comment]
@sedefsavas Thank you very much for sharing your insights. Currently there are only 2 additional versions to the default version available so i guess the cleanup works as expected .

I'll close the issue.

<sub>Matthias Lembcke <[matthias.lembcke@mercedes-benz.com](mailto:matthias.lembcke@mercedes-benz.com)>, Daimler TSS GmbH ([Provider Information](https://github.com/mercedes-benz/daimler-foss/blob/master/PROVIDER_INFORMATION.md))</sub>
